packages/core/src/libraries/domain.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  type CloudflareData,
+  type Domain,
+  type DomainDnsRecords,
+  DomainStatus,
+} from '@logto/schemas';
+
+import type Queries from '#src/tenants/Queries.js';
+import SystemContext from '#src/tenants/SystemContext.js';
+import assertThat from '#src/utils/assert-that.js';
+import { getCustomHostname, createCustomHostname } from '#src/utils/cloudflare/index.js';
+
+export type DomainLibrary = ReturnType<typeof createDomainLibrary>;
+
+const getDomainStatusFromCloudflareData = (data: CloudflareData): DomainStatus => {
+  if (data.status === 'active' && data.ssl.status === 'active') {
+    return DomainStatus.Active;
+  }
+
+  if (data.status === 'active' && data.ssl.status !== 'active') {
+    return DomainStatus.PendingSsl;
+  }
+
+  if (data.status !== 'active') {
+    return DomainStatus.PendingVerification;
+  }
+
+  return DomainStatus.Error;
+};
+
+export const createDomainLibrary = (queries: Queries) => {
+  const {
+    domains: { updateDomainById },
+  } = queries;
+
+  const syncDomainStatusFromCloudflareData = async (
+    domain: Domain,
+    cloudflareData: CloudflareData,
+    origin: string
+  ): Promise<Domain> => {
+    const status = getDomainStatusFromCloudflareData(cloudflareData);
+
+    const verificationError =
+      cloudflareData.verification_errors && cloudflareData.verification_errors.length > 0
+        ? cloudflareData.verification_errors.join('\n')
+        : null;
+    const sslError =
+      cloudflareData.ssl.validation_errors && cloudflareData.ssl.validation_errors.length > 0
+        ? cloudflareData.ssl.validation_errors.map(({ message }) => message).join('\n')
+        : null;
+    const errorMessage = [verificationError, sslError].filter(Boolean).join('\n');
```

Comment:
It seems we can first concat the two messages and then joining by `\n`. E.g.
```ts
const errorMessage = [
  ...cloudflareData.verification_errors,
  ...cloudflareData.map(({ message }) => message),
].filter(Boolean).join(\n);
```

Reply:
updated

Reply:
```suggestion
    const errorMessage = [...verificationError, ...sslError].filter(Boolean).join('\n');
```

is this not covered by tests?

Reply:
@gao-sun He joined the array first, so the `verificationError` and `sslError` in the array are strings now. But it could've just spreaded the two arrays into a new array and join by `\n` after that

Reply:
Yes, this is coverd by unit tests

Reply:
just personal preference, would prefer join in the end. but either way